hugr-model/src/v0/mod.rs

```diff
@@ -662,40 +670,42 @@ pub enum Term<'a> {
     ///
     /// `ctrl : static`
     ControlType,
+
+    /// Constraint that requires a runtime type to be copyable.
+    CopyConstraint {
+        /// The runtime type that must be copyable.
+        term: TermId,
+    },
+
+    /// Constraint that requires a runtime type to be discardable.
+    DiscardConstraint {
+        /// The runtime type that must be discardable.
+        term: TermId,
+    },
 }
 
 /// A parameter to a function or alias.
 ///
 /// Parameter names must be unique within a parameter list.
 /// Implicit and explicit parameters share a namespace.
 #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub enum Param<'a> {
-    /// An implicit parameter that should be inferred, unless a full application form is used
+pub struct Param<'a> {
```

Comment:
+1 to this commoning up, amen.

I am a long way from convinced that we need implicit/explicit, but that's another issue and definitely not for this PR!

Reply:
(nice that Constraints are now separate from Parameters here too! 👍  )